File: dsa/solution/120_Triangle.cpp
#include <iostream>
#include <vector>

using namespace std;
// Top-Down DP or Memoization
class Solution
{
public:
    int dfs(int i, int j, int n, vector<vector<int>> &triangle, vector<vector<int>> &memo)
    {
        if (i == n)
            return 0;
        if (memo[i][j] != -1)
            return memo[i][j];

        int lower_left = triangle[i][j] + dfs(i + 1, j, n, triangle, memo);
        int lower_right = triangle[i][j] + dfs(i + 1, j + 1, n, triangle, memo);
        memo[i][j] = min(lower_left, lower_right);

        return memo[i][j];
    }

    int minimumTotal(vector<vector<int>> &triangle)
    {
        int n = triangle.size();
        vector<vector<int>> memo(n, vector<int>(n, -1));
        return dfs(0, 0, n, triangle, memo);
    }
};

int main()
{
    Solution solution;

    // Example test case
    vector<vector<int>> triangle = {
        {2},
        {3, 4},
        {6, 5, 7},
        {4, 1, 8, 3}
    };

    int result = solution.minimumTotal(triangle);

    cout << "Minimum path sum: " << result << endl;

    return 0;
}
/*
    Time Complexity: O(n^2)
    Space Complexity: O(n^2)
*/
    

File: dsa/solution/120_Triangle.js
// Bottom Up DP or Tabulation
class Solution {
    minimumTotal(triangle) {
        let n = triangle.length;
        let dp = new Array(n).fill(0).map(() => new Array(n).fill(-1));
        dp[n - 1] = [...triangle[n - 1]];

        for (let i = n - 2; i >= 0; i--) {
            for (let j = 0; j <= i; j++) {
                let lowerLeft = triangle[i][j] + dp[i + 1][j];
                let lowerRight = triangle[i][j] + dp[i + 1][j + 1];
                dp[i][j] = Math.min(lowerLeft, lowerRight);
            }
        }

        return dp[0][0];
    }
}

// Example usage:
const solution = new Solution();
const triangle = [
    [2],
    [3, 4],
    [6, 5, 7],
    [4, 1, 8, 3]
];
const result = solution.minimumTotal(triangle);
console.log(result);
/*
    Time Complexity: O(n^2)
    Space Complexity: O(n^2)
*/

